editor/store/effects.js

```diff
@@ -154,14 +157,61 @@ export default {
 
 		request.then(
 			( newPost ) => {
-				const reset = isAutosave ? resetAutosave : resetPost;
-				dispatch( reset( newPost ) );
+				// An autosave may be processed by the server as a regular save
+				// when its update is requested by the author and the post was
+				// draft or auto-draft.
+				const isRevision = newPost.id !== post.id;
+
+				// Thus, the following behaviors occur:
+				//
+				//  - If it was a revision, it is treated as latest autosave
+				//    and updates optimistically applied are reverted.
+				//  - If it was an autosave but not revision under the above
+				//    noted conditions, cherry-pick updated properties since
+				//    the received revision entity shares some but not all
+				//    properties of a post.
+				//  - Otherwise, it was a full save and the received entity
+				//    can be considered the new reset post.
+				let updateAction;
+				if ( isRevision ) {
+					updateAction = resetAutosave( newPost );
+				} else if ( isAutosave ) {
+					const revisionEdits = pick( newPost, [
+						// Autosave content fields.
+						'title',
+						'content',
+						'excerpt',
+
+						// UI considers save to have occurred if modified date
+						// of post changes (e.g. PostPreviewButton).
+						//
+						// TODO: Consider formalized pattern for identifying a
+						// save as having completed.
+						'date',
+						'date_gmt',
+						'modified',
+						'modified_gmt',
+					] );
+
+					updateAction = updatePost( revisionEdits );
+				} else {
+					updateAction = resetPost( newPost );
+				}
+
+				dispatch( updateAction );
 
 				dispatch( {
 					type: 'REQUEST_POST_UPDATE_SUCCESS',
 					previousPost: post,
 					post: newPost,
-					optimist: { type: COMMIT, id: POST_UPDATE_TRANSACTION_ID },
+					optimist: {
+						// Note: REVERT is not a failure case here. Rather, it
+						// is simply reversing the assumption that the updates
+						// were applied to the post proper, such that the post
+						// treated as having unsaved changes.
+						type: isRevision ? REVERT : COMMIT,
```

Comment:
Instead of a revert would the following idea work:?
Save the post the autosave logic (but send flag) and maybe pick the fields to send, and then when we receive the return value from the server we compare the fields against the edits. In the case, a field was present in the edits and not returned by the server we use an action e.g: RESTORE_EDITS to restore those edits. This would make the post dirty again because some fields edits are not saved.

Reply:
>Instead of a revert would the following idea work:?

This reminds me of another goal: For autosaving a new draft post, ideally the user should not be prompted about unsaved changes after the autosave completes. This only works if we know that all fields are transmitted and considered in the save request. 

Rather than a standalone action, something like the `isChanging` implementation in [my comment above](https://github.com/WordPress/gutenberg/pull/7130#issuecomment-394838613) might serve the same purpose to what you describe.